File: dart_sample/lib/demos/general/focus_traversal_demo_page.dart
import 'package:material_ui/material_ui.dart';

import '../../counter_widgets.dart';

enum FocusTraversalDemoPolicy { readingOrder, widgetOrder, ordered }

class FocusTraversalDemoPage extends StatefulWidget {
  const FocusTraversalDemoPage({super.key});

  @override
  State<FocusTraversalDemoPage> createState() => _FocusTraversalDemoPageState();
}

class _FocusTraversalDemoPageState extends State<FocusTraversalDemoPage> {
  static const List<String> tileLabels = <String>['A', 'B', 'C', 'D', 'E', 'F'];

  final List<FocusNode> _nodes = <FocusNode>[];
  FocusTraversalDemoPolicy _policy = FocusTraversalDemoPolicy.readingOrder;
  String _focused = 'none';

  @override
  void initState() {
    super.initState();
    for (final String label in tileLabels) {
      _nodes.add(FocusNode(debugLabel: label));
    }
  }

  @override
  void dispose() {
    for (final FocusNode node in _nodes) {
      node.dispose();
    }
    _nodes.clear();
    super.dispose();
  }

  String get _policyLabel => switch (_policy) {
    FocusTraversalDemoPolicy.widgetOrder => 'Policy: widget order',
    FocusTraversalDemoPolicy.ordered => 'Policy: numeric order',
    FocusTraversalDemoPolicy.readingOrder => 'Policy: reading order',
  };

  FocusTraversalPolicy _createPolicy() => switch (_policy) {
    FocusTraversalDemoPolicy.widgetOrder => WidgetOrderTraversalPolicy(),
    FocusTraversalDemoPolicy.ordered => OrderedTraversalPolicy(),
    FocusTraversalDemoPolicy.readingOrder => ReadingOrderTraversalPolicy(),
  };

  void _cyclePolicy() {
    setState(() {
      _policy = switch (_policy) {
        FocusTraversalDemoPolicy.readingOrder => FocusTraversalDemoPolicy.widgetOrder,
        FocusTraversalDemoPolicy.widgetOrder => FocusTraversalDemoPolicy.ordered,
        FocusTraversalDemoPolicy.ordered => FocusTraversalDemoPolicy.readingOrder,
      };
    });
  }

  void _handleFocusChange(String label, bool focused) {
    setState(() {
      if (focused) {
        _focused = label;
      } else if (_focused == label) {
        _focused = 'none';
      }
    });
  }

  Widget _tile(int index) {
    final String label = tileLabels[index];
    final FocusNode node = _nodes[index];
    Widget tile = Expanded(
      child: Focus(
        focusNode: node,
        autofocus: index == 0,
        onFocusChange: (bool focused) => _handleFocusChange(label, focused),
        child: GestureDetector(
          onTap: node.requestFocus,
          child: Container(
            color: node.hasPrimaryFocus ? Colors.blue : Colors.grey.shade300,
            padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 18),
            child: Text(
              label,
              textAlign: TextAlign.center,
              style: TextStyle(
                fontSize: 16,
                color: node.hasPrimaryFocus ? Colors.white : Colors.black,
              ),
            ),
          ),
        ),
      ),
    );

    // Tile E stays focusable by tap or by the arrow keys, but Tab skips it.
    if (label == 'E') {
      tile = ExcludeFocusTraversal(child: tile);
    }

    // The ordered policy sorts the bottom row before the top one.
    return FocusTraversalOrder(
      order: NumericFocusOrder(index < 3 ? index + 10 : (index - 3).toDouble()),
      child: tile,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'FocusTraversalGroup + policies',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Tab and Shift+Tab walk the sorted order; the arrow keys use the geometric '
          'directional policy. Tile E is excluded from traversal but stays focusable.',
          style: TextStyle(fontSize: 14, color: Colors.grey),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            Expanded(
              child: CounterTapButton(
                label: _policyLabel,
                onTap: _cyclePolicy,
                background: Colors.blue,
                foreground: Colors.white,
                fontSize: 13,
              ),
            ),
            Expanded(
              child: CounterTapButton(
                label: 'Next',
                onTap: () => FocusManager.instance.primaryFocus?.nextFocus(),
                background: Colors.green,
                foreground: Colors.white,
                fontSize: 13,
              ),
            ),
            Expanded(
              child: CounterTapButton(
                label: 'Previous',
                onTap: () => FocusManager.instance.primaryFocus?.previousFocus(),
                background: Colors.blueGrey,
                foreground: Colors.white,
                fontSize: 13,
              ),
            ),
          ],
        ),
        FocusTraversalGroup(
          policy: _createPolicy(),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            spacing: 8,
            children: <Widget>[
              Row(spacing: 8, children: <Widget>[_tile(0), _tile(1), _tile(2)]),
              Row(spacing: 8, children: <Widget>[_tile(3), _tile(4), _tile(5)]),
            ],
          ),
        ),
        Text(
          'Focused tile: $_focused',
          style: const TextStyle(fontSize: 14, color: Colors.black),
        ),
      ],
    );
  }
}

File: src/Sample/Plumix.Sample/Demos/General/FocusTraversalDemoPage.cs
using System;
using System.Collections.Generic;
using Avalonia;
using Avalonia.Media;
using Plumix.Foundation;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/general/focus_traversal_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class FocusTraversalDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new FocusTraversalDemoPageState();
    }
}

internal enum FocusTraversalDemoPolicy
{
    ReadingOrder,
    WidgetOrder,
    Ordered,
}

internal sealed class FocusTraversalDemoPageState : State
{
    private static readonly string[] TileLabels = ["A", "B", "C", "D", "E", "F"];

    private readonly List<FocusNode> _nodes = [];
    private FocusTraversalDemoPolicy _policy = FocusTraversalDemoPolicy.ReadingOrder;
    private string _focused = "none";

    public override void InitState()
    {
        base.InitState();
        foreach (string label in TileLabels)
        {
            var node = new FocusNode(debugLabel: label);
            _nodes.Add(node);
        }
    }

    public override void Dispose()
    {
        foreach (FocusNode node in _nodes)
        {
            node.Dispose();
        }

        _nodes.Clear();
        base.Dispose();
    }

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("FocusTraversalGroup + policies", fontSize: 20, color: Colors.Black),
                new Text(
                    "Tab and Shift+Tab walk the sorted order; the arrow keys use the geometric "
                    + "directional policy. Tile E is excluded from traversal but stays focusable.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        new Expanded(
                            child: new CounterTapButton(
                                label: PolicyLabel,
                                onTap: CyclePolicy,
                                background: Colors.SteelBlue,
                                foreground: Colors.White,
                                fontSize: 13)),
                        new Expanded(
                            child: new CounterTapButton(
                                label: "Next",
                                onTap: () => _ = FocusManager.Instance.PrimaryFocus?.NextFocus(),
                                background: Colors.SeaGreen,
                                foreground: Colors.White,
                                fontSize: 13)),
                        new Expanded(
                            child: new CounterTapButton(
                                label: "Previous",
                                onTap: () => _ = FocusManager.Instance.PrimaryFocus?.PreviousFocus(),
                                background: Colors.SlateGray,
                                foreground: Colors.White,
                                fontSize: 13)),
                    ]),
                new FocusTraversalGroup(
                    policy: CreatePolicy(),
                    child: new Column(
                        crossAxisAlignment: CrossAxisAlignment.Stretch,
                        spacing: 8,
                        children:
                        [
                            new Row(spacing: 8, children: [Tile(0), Tile(1), Tile(2)]),
                            new Row(spacing: 8, children: [Tile(3), Tile(4), Tile(5)]),
                        ])),
                new Text($"Focused tile: {_focused}", fontSize: 14, color: Colors.Black),
            ]);
    }

    private string PolicyLabel => _policy switch
    {
        FocusTraversalDemoPolicy.WidgetOrder => "Policy: widget order",
        FocusTraversalDemoPolicy.Ordered => "Policy: numeric order",
        _ => "Policy: reading order",
    };

    private FocusTraversalPolicy CreatePolicy() => _policy switch
    {
        FocusTraversalDemoPolicy.WidgetOrder => new WidgetOrderTraversalPolicy(),
        FocusTraversalDemoPolicy.Ordered => new OrderedTraversalPolicy(),
        _ => new ReadingOrderTraversalPolicy(),
    };

    private void CyclePolicy()
    {
        SetState(() =>
        {
            _policy = _policy switch
            {
                FocusTraversalDemoPolicy.ReadingOrder => FocusTraversalDemoPolicy.WidgetOrder,
                FocusTraversalDemoPolicy.WidgetOrder => FocusTraversalDemoPolicy.Ordered,
                _ => FocusTraversalDemoPolicy.ReadingOrder,
            };
        });
    }

    private Widget Tile(int index)
    {
        string label = TileLabels[index];
        FocusNode node = _nodes[index];
        Widget tile = new Expanded(
            child: new Focus(
                focusNode: node,
                autofocus: index == 0,
                onFocusChange: focused => HandleFocusChange(label, focused),
                child: new GestureDetector(
                    onTap: () => node.RequestFocus(),
                    child: new Container(
                        color: node.HasPrimaryFocus ? Colors.SteelBlue : Colors.Gainsboro,
                        padding: new Thickness(14, 18),
                        child: new Text(
                            label,
                            fontSize: 16,
                            color: node.HasPrimaryFocus ? Colors.White : Colors.Black,
                            textAlign: TextAlign.Center)))));

        // Tile E stays focusable by tap or by the arrow keys, but Tab skips it.
        if (label == "E")
        {
            tile = new ExcludeFocusTraversal(child: tile);
        }

        // The ordered policy sorts the bottom row before the top one.
        return new FocusTraversalOrder(
            order: new NumericFocusOrder(index < 3 ? index + 10 : index - 3),
            child: tile);
    }

    private void HandleFocusChange(string label, bool focused)
    {
        SetState(() =>
        {
            if (focused)
            {
                _focused = label;
            }
            else if (_focused == label)
            {
                _focused = "none";
            }
        });
    }
}
